include/sys/queued_operation.h

```diff
@@ -0,0 +1,239 @@
+	const struct queued_operation_functions *vtable;
+
+	/* Lock controlling access to other fields. */
+	struct k_spinlock lock;
+
+	/* The operation that is being processed. */
+	struct queued_operation *current;
+
+	/* A flag indicating that the last operation has been
+	 * completed and the callback notification is being
+	 * executed.
+	 */
+	bool finalizing;
+};
+
+#define QUEUED_OPERATION_MANAGER_INITIALIZER(_vtable) {	\
+		.vtable = _vtable,			\
+}
+
+/**
+ * @brief Submit an operation to be processed when the service is
+ * available.
+ *
+ * During the call to this function the service process function will
+ * have been invoked at least once, either providing another operation
+ * or indicating that no operations are pending.
+ *
+ * @param mgr a generic pointer to the service instance
+ * @param op a generic pointer to an operation to be performed
+ * @param priority the priority of the operation relative to other
+ * operations.  Numerically lower values are higher priority.  Values
+ * outside the range of a signed 8-bit integer will be rejected.
+ *
+ * @retval -ENOTSUP if callback notification is requested and the
+ * service does not provide a callback translation.  This may also be
+ * returned due to service-specific validation.
+ *
+ * @retval -EINVAL if the passed priority is out of the range of
+ * supported priorities.  This may also be returned due to
+ * service-specific validation.
+ *
+ * @return A negative value if the operation was rejected by the
+ * service or due to other configuration errors.  A non-negative value
+ * indicates the operation has been accepted for processing and
+ * completion notification will be provided.
+ */
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority);
+
+/**
+ * @brief Helper to extract the result from a queued operation.
+ *
+ * This forwards to :cpp:func:`async_notify_fetch_result()`.
+ */
+static inline int queued_operation_fetch_result(struct queued_operation *op,
+						int *result)
+{
+	return async_notify_fetch_result(&op->notify, result);
+}
+
+/**
+ * @brief Attempt to cancel a queued operation.
+ *
+ * Successful cancellation issues a completion notification with
+ * result -ECANCELED for the submitted operation before this function
+ * returns.
+ *
+ * @retval 0 if successfully cancelled.
+ * @retval -EINPROGRESS if op is currently being executed, so cannot
+ * be cancelled.
+ * @retval -EINVAL if op is neither being executed nor in the queue of
+ * pending operations
+ */
+int queued_operation_cancel(struct queued_operation_manager *mgr,
+			    struct queued_operation *op);
+
+/**
+ * @brief Send the completion notification for a queued operation.
+ *
+ * This function must be invoked by services that support queued
+ * operations when the operation provided to them through the process
+ * function have been completed.  It is not intended to be invoked by
```

Comment:
maybe it's worth stressing that this should be called at the and of completion notification because it may result with process call with NULL operation before actual action is performed. For example if process function is synchronous then finalizing may look like:
```
void process(struct queued_operation_manager *mgr,
		struct queued_operation *op)
{
	if (op) {
		queued_operation_finalize(mgr, op, 0);
                // do some processing
	} else {
		// shut down resources
	}
}
```
In the case above process(mng, NULL) will be called in the context of `queued_operation_finalize` call, before actual action and that's probably unintended.

It is implicitly explained in the sentence "During the call..." but i wonder if it is enough.